### PR TITLE
fix(typescript): write json srcs to generated tsconfig for resolveJso…

### PR DESCRIPTION
**Pull request by matthewjh**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

When using a `ts_project` rule with a generated tsconfig file, tsc errors when using `resolveJsonModule`, because the generated tsconfg file does not enumerate the .json inputs in the "files" array. For reasons unclear to me, tsc only enforces this on `composite` projects.

## What is the new behavior?

.json inputs are enumerated in generated tsconfig files.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/typescript/internal/ts_project.bzl [hunk 1]
```diff
@@ -323,6 +323,9 @@ def _is_ts_src(src, allow_js):
         return True
     return allow_js and (src.endswith(".js") or src.endswith(".jsx"))
 
+def _is_json_src(src):
+    return src.endswith(".json")
```
Comment (matthewjh) [inline packages/typescript/internal/ts_project.bzl:327]: I am including JSON srcs unconditionally. This _may_ be fine as `tsc` will error if JSON files are imported without `resolveJsonModule`.

On the other hand, they could be included conditionally on `resolveJsonModule` being set to `True` either in an attribute on the rule or nested in the tsconfig.compilerOptions map. This would be similar to how `allowJs` drives the inclusion of JS files.

Comment (matthewjh) [reply packages/typescript/internal/ts_project.bzl:327]: I think unconditional inclusion may cause a problem if people are globing `**/*` and this sucks up incidental JSON files. I'm not sure how `tsc` behaves if JSON files are included in its compilation without being imported from any TS files. Whatever the case, it may become more or less lenient in future, so maybe we should play safe here.

Any thoughts?

Comment (alexeagle) [reply packages/typescript/internal/ts_project.bzl:327]: We definitely need to handle those corner cases correctly. If you're not sure how `tsc` behaves then do that experiment, or add that scenario under packages/typescript/test/ts_project and let's verify?

Comment (matthewjh) [reply packages/typescript/internal/ts_project.bzl]: ok, should be good now. I went down the route of:

1. adding a new attribute to the `ts_project` macro & rule, `resolve_json_module`
2. only including json srcs if `resolve_json_module` is true, as happens for `allow_js`
3. auto-including json srcs if `srcs` is None and `resolve_json_module` is true, as happens for `allow_js`

I did notice some inconsistency as to how some of the attributes are handled versus the equivalent inside `ts_config.compilerOptions`, but that is more general than this so I'd like to probe that in a separate PR.

File: packages/typescript/internal/ts_project.bzl [hunk 2]
```diff
@@ -610,10 +614,14 @@ def ts_project_macro(
     """
 
     if srcs == None:
+        include = ["**/*.ts", "**/*.tsx", "**/*.js", "**/*.jsx"]
```
Comment (alexeagle) [inline packages/typescript/internal/ts_project.bzl:617]: something is wrong with the logic here, you can never leave this block without including .js and .jsx where previously you could

File: packages/typescript/internal/ts_project.bzl [hunk 3]
```diff
@@ -629,6 +637,7 @@ def ts_project_macro(
         declaration_map = compiler_options.setdefault("declarationMap", declaration_map)
         emit_declaration_only = compiler_options.setdefault("emitDeclarationOnly", emit_declaration_only)
         allow_js = compiler_options.setdefault("allowJs", allow_js)
+        resolve_json_module = compiler_options.setdefault("resolveJsonModule", resolve_json_module)
```
Comment (alexeagle) [inline packages/typescript/internal/ts_project.bzl:640]: you should also update the validator to check that the attribute setting matches what's in the tsconfig file

**Review by alexeagle — COMMENTED** [review]
thanks!

File: packages/typescript/internal/ts_project.bzl (hunk 3 above)
Comment (matthewjh) [reply packages/typescript/internal/ts_project.bzl:640]: thanks, fixed. are there any tests for the validator? i couldn't find any.

File: packages/typescript/internal/ts_project.bzl (hunk 2 above)
Comment (matthewjh) [reply packages/typescript/internal/ts_project.bzl]: good spot, thanks. i feel that internal details like this would be easier to catch with unit tests, so we don't have to figure out ways to make everything express through mini-integration tests (which is hard for this one, for instance). but i guess the tooling is not in place for this?

Comment (matthewjh) [reply packages/typescript/internal/ts_project.bzl]: `generated_file_tests` for generated tsconfigs would have caught this. that's not a substitute for testing some of these internal pieces separately, but would be an improvement on what's there today. perhaps i can look at that later on in a different pr.

**Comment by matthewjh** [comment]
@alexeagle this should be ready now.

File: packages/typescript/internal/ts_project.bzl (hunk 2 above)
Comment (alexeagle) [reply packages/typescript/internal/ts_project.bzl]: yes sometimes the "mini-integration test" is not the best choice. There's a way to assert on what actions are produced by a rule implementation but not run them (like running `bazel aquery` and asserting on the result" which I'd love to try.

maybe for this even a smaller test fixture using the testing libraries in skylib would work, but I haven't investigated those for a while.

File: packages/typescript/internal/ts_project.bzl (hunk 3 above)
Comment (alexeagle) [reply packages/typescript/internal/ts_project.bzl:641]: not for the negative case, no

**Review by alexeagle — APPROVED** [review]
thanks!
